### PR TITLE
feat: follow research articles for updates #1438

### LINKED ISSUE #1438: [feature request] Subscribe to Research pieces
**Is your feature request related to a problem? Please describe.**

Research articles are intended to be an ongoing collection of updates around a single theme. The research themes will be documented during the process and will unfold over weeks, months and years. 

Ideally people are able to engage with the research process at an early stage and follow it's progress. 

**Describe the solution you'd like**

The option to subscribe to a Research post so that users could receive notifications as updates are made available. This could be done with one or both of the following options:

 * Platform notifications: we recently introduced the ability for Users to receive notifications around Useful Votes/Comments. 
* RSS feeds. This mechanism would allow readers to integrate it into their own workflow. The technology has fallen into disuse over the last few years due to gaps in the underlying technology which make it hard to monetise. Luckily that is not a metric which drives our decision making.

### PR DESCRIPTION
**Pull request by thisislawatts**
PR Checklist

- [X] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

- feat(components): add support for research_update notification
- feat: follow research article

Outstanding;
- [ ] Integration test for ResearchDescription.
- [x] Unsubscribe from article, logic exists in store but not hooked up to button
- [x] Toggle button status to show if you are already following
- [ ] Design review

## Git Issues

Closes #1438

Slight deviation from original designs to align with new Useful button styling. 
Screenshot 2023-03-07 at 18 38 34

### CODE REVIEW

**Comment by davehakkens** [comment]
> Sanity check, a user probably doesn't want to subscribe to **any** updates to an article. What constitutes a meaningful change?

I would say:
Subscribe to: any new update (step) posted.
Not subscribe to: any new comment made

**Comment by davehakkens** [comment]
Nice one. Seems promosing!
But I dont see any follow button?
afbeelding

**Comment by thisislawatts** [comment]
Ah weird, I am seeing it locally but not in the preview 🤔 let me take a look…

**Comment by davehakkens** [comment]
Did some local testing.
1. Button to toggle works, but when following I dont receive a notification if an update is posted.
2. Can we make the icon `orange` (like notifications) when its active. To make the different state is more visable

**Comment by thisislawatts** [comment]
> 1. Button to toggle works, but when following I dont receive a notification if an update is posted.
Will take another look into this 👀 
> 2. Can we make the icon `orange` (like notifications) when its active. To make the different state is more visable
Sure thing!

**Comment by thisislawatts** [comment]
After some further investigation @davehakkens can we focus on merging the changes which introduce the core functionality and iterate on the Following Button in a separate PR? 

This Following Button will only be available for `beta-tester` users. 

I am wondering whether we should align with the Useful button and show how many people are Following a Research project.

**Comment by davehakkens** [comment]
Works now. Nice. Yeh i'd buy up for releasing and iterating as we go.

**Comment by thisislawatts** [comment]
Okay great, thanks @davehakkens. With the button being worked on separately, @ONEARMY/maintainers can I get a review for the functional changes here?

**Comment by AlfonsoGhislieri** [comment]
I'll get this reviewed today!

File: src/stores/Research/research.store.test.ts [hunk 1]
```diff
@@ -620,4 +631,87 @@ describe('research.store', () => {
...
+          subscribers: ['existing-subscriber'],
+        },
+      )
+
+      // Act
+      await store.addSubscriberToResearchArticle(
+        researchItem._id,
+        'an-interested-user',
+      )
+
```
Comment (AlfonsoGhislieri) [inline src/stores/Research/research.store.test.ts:648]: **Nitpick** Could you add `// Assert `for better readability

File: src/stores/Research/research.store.test.ts [hunk 2]
```diff
@@ -620,4 +631,87 @@ describe('research.store', () => {
...
+          subscribers: ['long-term-subscriber', 'remove-me'],
+        },
+      )
+
+      // Act
+      await store.removeSubscriberFromResearchArticle(
+        researchItem._id,
+        'remove-me',
+      )
+
```
Comment (AlfonsoGhislieri) [inline src/stores/Research/research.store.test.ts:686]: Same as above

File: src/stores/Research/research.store.test.ts [hunk 3]
```diff
@@ -620,4 +631,87 @@ describe('research.store', () => {
...
+    it('triggers a notification for each subscribed users', async () => {
+      const { store, researchItem, setFn } = await factoryResearchItem({
+        subscribers: ['subscriber'],
+      })
+
+      // Act
+      await store.uploadUpdate(FactoryResearchItemUpdate())
+      await store.uploadUpdate(FactoryResearchItemUpdate())
+
```
Comment (AlfonsoGhislieri) [inline src/stores/Research/research.store.test.ts:704]: Also same as above

File: src/stores/Research/research.store.tsx [hunk 4]
```diff
@@ -135,6 +135,52 @@ export class ResearchStore extends ModuleStore {
...
+    if (researchData && !(researchData?.subscribers || []).includes(userId)) {
+      await this.updateResearchItem(dbRef, {
+        ...researchData,
+        subscribers: [userId].concat(researchData?.subscribers || []),
+      })
+
+      const createdItem = (await dbRef.get()) as IResearch.ItemDB
+      runInAction(() => {
+        this.activeResearchItem = createdItem
+      })
```
Comment (AlfonsoGhislieri) [inline src/stores/Research/research.store.tsx:154]: **Question - curious** Is this line to update the `activeReseachItem` that is observable to other components?

File: src/stores/Research/research.store.tsx [hunk 5]
```diff
@@ -523,6 +571,31 @@ export class ResearchStore extends ModuleStore {
...
+    const subscribers = researchItem.subscribers || []
+
+    // Only notify subscribers if there is a new update added
+    logger.debug('Notify each subscriber', {
+      subscribers,
+      beforeUpdateNumber: previousVersion?.updates
+        ? previousVersion?.updates.length
+        : 0,
+      afterUpdateNumber: researchItem?.updates.length,
+    })
```
Comment (AlfonsoGhislieri) [inline src/stores/Research/research.store.tsx:584]: **Question** Was this kept in for future potential debugging?

**Review by AlfonsoGhislieri — CHANGES_REQUESTED** [review]

**Comment by AlfonsoGhislieri** [comment]
All done @thisislawatts! Just a few questions and small nitpicky changes, but the main issue that I found was not handling a user that is not logged in, this is obviously not an issue right now since this is only available to `beta-testers` who have to be logged in. So not sure if this should be addressed at this time!

File: src/models/research.models.tsx [hunk 6]
```diff
@@ -12,6 +12,8 @@ export type IResearchStats = {
   votedUsefulCount: number
 }
 
+type UserIdList = string[]
```
[the remark this answers is not in the document]
Comment (thisislawatts) [reply src/models/research.models.tsx:15]: You're right that we don't get much additional value from this type. The intention here is to clearly communicate the type of strings which will be contained within this list. Happy to remove if you think this additional type does not make the code easier to understand.

File: packages/components/src/NotificationItem/NotificationItem.tsx [hunk 7]
```diff
@@ -19,6 +20,8 @@ export interface UserNotificationItem {
 function getIconByType(type: notificationType): availableGlyphs {
   return ['howto_useful', 'research_useful'].includes(type)
     ? 'useful'
+    : type === 'research_update'
+    ? 'thunderbolt'
```
[the remark this answers is not in the document]
Comment (thisislawatts) [reply packages/components/src/NotificationItem/NotificationItem.tsx]: Added here https://github.com/ONEARMY/community-platform/pull/2145/commits/e88cbbb9df474b4514c5ec4ca59c71c90bce7e07

File: src/pages/Research/Content/ResearchDescription.tsx [hunk 8]
```diff
@@ -105,6 +106,26 @@ const ResearchDescription = ({ research, isEditable, ...props }: IProps) => {
...
+              onClick={() => {
+                props.onFollowingClick()
+              }}
+            >
+              {research.subscribers?.includes(
+                props?.loggedInUser?.userName || '',
+              )
+                ? 'Following'
+                : 'Follow'}
+            </Button>
```
[the remark this answers is not in the document]
Comment (thisislawatts) [reply src/pages/Research/Content/ResearchDescription.tsx:127]: Agreed, I think this entire chunk needs to be refactored, at the moment it is wrapped in `AuthWrapper` which requires a logged in user. 

I suggest that this work should be done as part of #2191

File: src/stores/Research/research.store.tsx (hunk 4 above)
Comment (thisislawatts) [reply src/stores/Research/research.store.tsx:154]: Thats the intention here, I copied this pattern from elsewhere in the store. However thinking about it I think each of these should be replaced with a call to `setActiveResearchItem` so reduce duplicated mechanisms for update the activeResearchItem. 🤔 

Maybe a good item for a follow up refactoring PR

File: src/stores/Research/research.store.tsx (hunk 5 above)
Comment (thisislawatts) [reply src/stores/Research/research.store.tsx:584]: That was the intention.

File: src/pages/Research/Content/ResearchDescription.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlfonsoGhislieri) [reply src/pages/Research/Content/ResearchDescription.tsx:127]: Ah yeah that's definitely getting quite cluttered, thanks for raising the issue :-)

File: src/stores/Research/research.store.tsx (hunk 4 above)
Comment (AlfonsoGhislieri) [reply src/stores/Research/research.store.tsx:154]: Yeah I kept seeing across the store, could be worth looking into like you said for some refactoring!

**Comment by thisislawatts** [comment]
My preference @AlfonsoGhislieri would be to see the Follow/Folllowing button handled separately for a few reasons:
* This PR is already quite old and touches a bunch of files other people may be working on, the sooner these changes are available on the default branch the less likely we are to encounter merge conflicts. 
* The opt in/out button is not coupled to the backend logic so does not need to go out in the same PR. The use of `beta-tester` here helps protect users from seeing work in progress.
* There needs to be some design iterations to the Follow/Following button which should not block the functional change we are making here.

**Comment by AlfonsoGhislieri** [comment]
Yeah that makes complete sense to me @thisislawatts! Didn't see that you had raised that PR already. Great work! 👍🏻

**Review by AlfonsoGhislieri — APPROVED** [review]

**Review by davehakkens — APPROVED** [review]

**Pull request merged**